simulator/recorder/recorder.go

```diff
@@ -0,0 +1,193 @@
+package recorder
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"golang.org/x/xerrors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+type Event string
+
+var (
+	Add    Event = "Add"
+	Update Event = "Update"
+	Delete Event = "Delete"
+)
+
+type Service struct {
+	client   dynamic.Interface
+	gvrs     []schema.GroupVersionResource
+	path     string
+	recordCh chan Record
+}
+
+type Record struct {
+	Time     time.Time                 `json:"time"`
+	Event    Event                     `json:"event"`
+	Resource unstructured.Unstructured `json:"resource"`
+}
+
+var DefaultGVRs = []schema.GroupVersionResource{
+	{Group: "", Version: "v1", Resource: "namespaces"},
+	{Group: "scheduling.k8s.io", Version: "v1", Resource: "priorityclasses"},
+	{Group: "storage.k8s.io", Version: "v1", Resource: "storageclasses"},
+	{Group: "", Version: "v1", Resource: "persistentvolumeclaims"},
+	{Group: "", Version: "v1", Resource: "nodes"},
+	{Group: "", Version: "v1", Resource: "persistentvolumes"},
+	{Group: "", Version: "v1", Resource: "pods"},
+}
+
+type Options struct {
+	GVRs []schema.GroupVersionResource
+	Path string
+}
+
+func New(client dynamic.Interface, options Options) *Service {
+	gvrs := DefaultGVRs
+	if options.GVRs != nil {
+		gvrs = options.GVRs
+	}
+
+	return &Service{
+		client:   client,
+		gvrs:     gvrs,
+		path:     options.Path,
+		recordCh: make(chan Record),
+	}
+}
+
+func (s *Service) Run(ctx context.Context) error {
+	infFact := dynamicinformer.NewFilteredDynamicSharedInformerFactory(s.client, 0, metav1.NamespaceAll, nil)
+	for _, gvr := range s.gvrs {
+		inf := infFact.ForResource(gvr).Informer()
+		_, err := inf.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { s.RecordEvent(obj, Add) },
+			UpdateFunc: func(_, obj interface{}) { s.RecordEvent(obj, Update) },
+			DeleteFunc: func(obj interface{}) { s.RecordEvent(obj, Delete) },
+		})
+		if err != nil {
+			return xerrors.Errorf("failed to add event handler: %w", err)
+		}
+		infFact.Start(ctx.Done())
+		infFact.WaitForCacheSync(ctx.Done())
+	}
+
+	err := s.backup()
+	if err != nil {
+		klog.Error("Failed to backup record: ", err)
+	}
+
+	go s.record(ctx)
+
+	return nil
+}
+
+func (s *Service) RecordEvent(obj interface{}, e Event) {
+	unstructObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		klog.Error("Failed to convert runtime.Object to *unstructured.Unstructured")
+		return
+	}
+
+	r := Record{
+		Event:    e,
+		Time:     time.Now(),
+		Resource: *unstructObj,
+	}
+
+	s.recordCh <- r
+}
+
+func (s *Service) record(ctx context.Context) {
+	records := []Record{}
+	for {
+		select {
+		case r := <-s.recordCh:
```

Comment:
If I understand correctly, on every event (resource creation/update/delete) this method serializes all events that have happened so far and writes them to a file.
`recordCh` is unbuffered, so `RecordEvent` will be blocked until `record` completes handling an inflight event.

I guess we need some optimization like the following, otherwise the recorder would have a limited performance.

- Asynchronous: Change `recordCh` to a buffered channel so that `RecordEvent` will not be blocked.
- Buffering: Save some records in a buffer and write them together as a chunk to reduce file system operations
- Incremental: Write each record chunk into separate files to avoid writing all records many times.